docs/adr/adr-013-blocksync-overhaul-part-2.md

```diff
@@ -0,0 +1,418 @@
+EDS by retrieving and reconstructing it. As we're improving our EDS retrieval logic, the new __happy path__ over `ShrEx/EDS` 
+protocol will be integrated in there.
+
+
+
+## Hardening
+
+### Rate Limiting
+
+Both `ShrEx/EDS` and `ShrEx/ND` has a server side that serves data. To prevent trivial DOS vector rate limiting should
+be introduced. Both protocols should have a configurable amount of data requests they serve at any given moment. If any 
+new request exceeds the limit, the server responds with `REFUSED` status code to client, so that client can act accordingly
+and try another peer.
+
+The alternative is to queue clients' requests and make client wait until it's request can be processed.
+
+## Alternative approaches
+
+### Historical EDSes
+
+- Extract Bitswap engine with custom transport for EDSes instead of the IPFS blocks
+  - Solves both recent and historical EDS retrieval
+  - Uncertain amount of efforts and thus time
+
+### Recent EDS
+
+- Polling connected peers
+- `ExtendedHeader` gossip as acknowledgment that data is received and ready to be served.
+- Implement custom engine similar to Bitswap with WANTLIST semantics(WANT_HAVE, WANT_DATA, DONT_HAVE, DONT_WANT)
+  - Generic protocol that solves retrieval of historical and recent EDSes
+  - Does not fit in the current time budget
+
+[go-libp2p]: https://github.com/libp2p/go-libp2p
+[proto]: https://developers.google.com/protocol-buffers
+[backpressure-quic]: https://datatracker.ietf.org/doc/html/draft-ietf-quic-transport-19#section-4
+
+## ToBeRemoved/Refactored
+
+### Recent EDS
+
+There is a crucial edge-case for the recent data. When a Node receives a recent header it immediately goes to request the
+data via `das.Daser` from the network through `share.Availability` interface. However, there is no guarantee that upon the
+new `ExtendedHeader` any of the Node's peers will have the data downloaded and ready to be served. That is, gossiping
+of `ExtendedHeader` outpaces fetching of EDS data.
+
+To overcome the problem, we start using `FloodSub` router in addition to existing `GossipSub` router.
+
+- BNs __fan-out__(TODO Link) acknowledgement messages with the DataRoot(TODO Link or specify in glossary) hashes of the
+  newly produced EDS by the Core network. The acknowledgement means "I have data, request it from me".
+- FNs subscribe to these messages and apply further message processing
+  - Check that the received DataRoot hash is part of any recent `ExtendedHeader`
+  - Request the message sender peer(other BN/FN) for the ODS committed into the DataRoot.
+    - If the peer is lying or does not respond with the data it acknowledged to have - wait for acknowledges from other peers
+    - After certain period of time(TODO Define) fallback to polling discovered peers in the __full node__ subnetwork.
+  - Retransmit the message further to acknowledge connected peers.
```

Comment:
> only to directly connected peers? 

It's the goal, but we are not restricted to requesting only connected peers.

> but in real world will have diameter > 1
wdym by diameter? network degree?

> Did you consider to make FN request data from source of the message rather than from sender?

Yes, but we don't want to overload the sender. Imaging the whole network asking only one sender. But,` sender` and `source` is valid case in PubSub and thus in our logic. This means that with a small network, we will request the source, but requesting it from sender will decrease the load from sender, when network grows

Reply:
> wdym by diameter? network degree?

Sorry for confusion, I was referring to graph diameter. 

>  Imaging the whole network asking only one sender
I thought this issue is addressed by retransmitting only after obtaining EDS. If we consider rate-limiting approach by limiting amount of concurrent downloads, it means server side has fixed amount of slots for clients. So when FN(client) finishes download, the slot is released and becomes available. The FN could propagate info about this slot being available further to network by retransmitting same notification. And also sent notification about it being available as source as well. 

Actually if `floodSub` provides connection to more than half of the nodes in network, my suggestion won't provide any profit.